### ISSUE TITLE
for..of for ArrayLike

### ISSUE THREAD

**Issue by NN---** [Microsoft/TypeScript#22878]
I suggest to allow ArrayLike in for..of enumerations:
```ts
const a: ArrayLike<number> = [1, 2, 3];
for (const e of a) { ... }
```

The code with array works.
```ts
const a: Array<number> = [1, 2, 3];
for (const e of a) { ... }
```

It transpiles in ES3/5 to the following:
```js
var a = [1, 2, 3];
for (var _i = 0, a_1 = a; _i < a_1.length; _i++) {
    var e = a_1[_i];
}
```

The resulting code uses only length property and index property which are exactly ArrayLike interface:
```ts
interface ArrayLike<T> {
    readonly length: number;
    readonly [n: number]: T;
}
```

**Comment by NN---** [comment]
@RyanCavanaugh Any reason it doesn't work without downlevel iteration flag as it works for Array and ReadonlyArray ?

**Comment by RyanCavanaugh** [comment]
We don't know that an `ArrayLike` will necessarily iterate the same way with a `for` loop as a `for / of` loop. The flag is there as a "Use at your own risk" feature.

**Comment by NN---** [comment]
I get the point.
TS cannot express a type safe for for index loop and ArrayLike.
Thanks.

**Comment by NN---** [comment]
@RyanCavanaugh It seems that it is more confusing than I thought.
What is the reason to use Symbol.iterator for..of for arrays if for-index loop is sufficient ?
Is it something that I miss ?

In other words I wanted to preserve the simple loop which I get without 'downlevelIteration' with this flag enabled.

Source:
```ts
for (const e of array) {}
```

downlevelIteration: false:
```js
for (let i=0,n=array.length;i<n;i++) {
 const e = array[i];
 ...
}
```

downlevelIteration: true
```js
try {
 for (var array_1 = tslib_1.__values(array), array_1_1 = array_1.next(); !array_1_1.done; array_1_1 = array_1.next()) {
  .... 
  }
} catch (e) { 
  ...
} finally {
  ...
}
```
